### PR TITLE
Remove the dependency on astor for Pythons ≥ 3.9

### PR DESCRIPTION
**Pull request by Kodiologist**
Python 3.9 added `ast.unparse`, which can do most of the work. It has some bugs that I fixed in `astor`, which I've worked around here, but I've submitted a patch to CPython to fix them for real (python/cpython#24897).

### CODE REVIEW

File: hy/core/language.hy [hunk 1]
```diff
@@ -226,14 +226,16 @@
...
-       astor.code-gen.to-source
-       astor.dump-tree)
-    compiled))
+  (if
+    codegen
+      (hy._compat.ast-unparse compiled)
+    hy._compat.PY3_9
+      (ast.dump compiled :indent 1)
+    True
+      (ast.dump compiled)))
```
Comment (scauligi) [inline hy/core/language.hy:238]: Isn't the last form of an `if` already implicitly an `else`?

File: hy/_compat.py [hunk 2]
```diff
@@ -14,3 +15,33 @@ def reraise(exc_type, value, traceback=None):
...
+                    .replace("inf", ast._INFSTR)
+                    .replace('nan', '({}-{})'.format(ast._INFSTR, ast._INFSTR)))
+            else:
+                super().visit_Constant(node)
+
+        def visit_Set(self, node):
+            if node.elts:
+                super().visit_Set(node)
+            else:
+                self.write('{1}.__class__()')
```
Comment (scauligi) [inline hy/_compat.py:38]: (trivial, but) is there any significant difference between the representation here vs in your upstream patch `{*()}`?

**Review by peaceamongworlds — APPROVED** [review]

File: hy/core/language.hy (hunk 1 above)
Comment (Kodiologist) [reply hy/core/language.hy:238]: Yes, but I think it's easier to read with the placeholder `True`.

File: hy/_compat.py (hunk 2 above)
Comment (Kodiologist) [reply hy/_compat.py]: They should be equivalent. I've changed it to `{*()}`.

File: hy/_compat.py [hunk 3]
```diff
@@ -14,3 +15,33 @@ def reraise(exc_type, value, traceback=None):
         raise value.with_traceback(traceback)
     finally:
         traceback = None
+
+
+if PY3_9:
+
+    import ast, math, cmath
```
Comment (scauligi) [inline hy/_compat.py:22]: The `math` and `cmath` imports appear irrelevant.

File: hy/_compat.py [hunk 4]
```diff
@@ -14,3 +15,33 @@ def reraise(exc_type, value, traceback=None):
...
+    class Unparser(ast._Unparser):
+      # Work around some limitations of Python's `ast.unparse`.
```
Comment (scauligi) [inline hy/_compat.py:24]: Add a reference to https://github.com/python/cpython/pull/24897. Since it's going to be backported to 3.9, we can eventually remove this workaround once it lands.

**Review by scauligi — APPROVED** [review]
Aside from a few more small changes, lgtm

File: hy/_compat.py (hunk 3 above)
Comment (Kodiologist) [reply hy/_compat.py:22]: Yep, they're leftovers.

File: hy/_compat.py (hunk 4 above)
Comment (Kodiologist) [reply hy/_compat.py:24]: Done.
